libsolidity/analysis/SyntaxChecker.cpp

```diff
@@ -154,3 +176,13 @@ bool SyntaxChecker::visit(PlaceholderStatement const&)
 	return true;
 }
 
+bool SyntaxChecker::visit(MemberAccess const& _memberAccess)
+{
+	if (m_nonPayablePublic && !m_library) {
+		auto const& identifier = dynamic_cast<Identifier const&>(_memberAccess.expression());
```

Comment:
I think it is better to check the type of `_memberAccess.expression()`. If you do it like that, creating a struct called `msg` with a `value` member will also trigger the warning ( - please create a test for that case - ).
So you have to do something like the following:
```
if (MagicType const* t = dynamic_cast<MagicType const*>(_memberAccess.expression().annotation().type().get()) {
 // check the kind of magic and compare _memberAccess.memberName() against "value"
```
The problem is that you need to perform that after type checking, because only then, the type information will be available. I think the best solution would be to create another ast walker similar to syntax checker, that is run after type checking. That component could just be called `StaticAnalyzer`.